Defer error to first API call when config file references a missing context

simonferquel opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/cli/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

~~Previously, when the "CurrentContext" field of the config file referenced a context that did not exist in the store, the CLI would fail (even for simple things like docker context ls).~~
~~This causes an issue with Docker Desktop "Contexts Syncing" feature (that uses the same store between win32 and WSL2):~~
~~If win32 cli removes the "current context" of the WSL 2 CLI (which it can do if it is not the "current context" in the win32 config file), it would make the WSL 2 CLI unusable, asking the user to modify the config file by hand.~~

When a user removes a context from the store without using the CLI (by removing the context directory directly in the store), it can trigger an issue where the CLI does not work at all (even for changing the current context) anymore without editing the config file by hand.

With Docker Desktop and WSL 2, we now link the wsl 2 distros context store to the windows one, so that a context created from Windows can be used in WSL 2 and vice-versa. However we keep the CLI config files independant (at least for now), so this issue can be triggered quite often.

This commit fixes that by changing the way we handle context missing error: if the resolved context name comes from the config file and refers a missing context, the error is reported when the API is actually called instead of at the CLI initialization. Thus, local-only commands continue to work (shuch as `context use`, `context ls`,...).

~~This commit relaxes this behavior a bit: explicitly referencing a missing context (using the --context flag or DOCKER_CONTEXT env var) will still produce an error, but implicitly referencing such a missing context (via the CurrentContext config field) will fallback to use the "default" context.~~

**- How I did it**

~~In the function resolving the current context name, instead of erroring with asking the user to modify its config file when the config file references a missing context, fallback to default~~

The function resolving the current context name does not error when the config file references a missing context. It returns a hint about the source of context name.

The check for a missing context is done in the CLI init code. If we find out that the context is missing and that the source of the selected context name is the config file, instead of failing, we initialize the api client with an always failing client (that reports the error during dialing). This allows to change the current context easily, without having to edit the config file by hand.

**- How to verify it**

~~Edit the config file, set CurrentConfig to a value that is not an existing context name, check that the CLI fallbacks to using "default".~~

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on cli/command/cli.go:
```diff
@@ -549,7 +549,9 @@ func resolveContextName(opts *cliflags.CommonOptions, config *configfile.ConfigF
-			return "", errors.Errorf("Current context %q is not found on the file system, please check your config file at %s", config.CurrentContext, config.Filename)
+			// If config file references a context that does not exist, fallback to default
+			// Explicitly requested contexts (flag or env-var based) will still get the error
+			return DefaultContextName, nil
```
I don't think we should silently fall back, as this would lead to unexpected results (similar to outlined in https://github.com/docker/cli/issues/1637)

simonferquel replied:
Would writing a message to stderr feel good enough ?

thaJeztah replied:
No, I don't think a message would be enough, because that wouldn't make the command fail.

Consider a situation where I am in a specific context, and that context is (somehow) removed, and all of a sudden my command completes successfully but against the wrong server. Depending on the command I ran that could mean I just (e.g.)  _removed_ a volume or image from the wrong environment.

simonferquel replied:
Indeed. Would deferring the error to the http call be enough?

thaJeztah replied:
If we can defer to the moment it's actually needed, yes, that'd work (probably also when doing `docker context inspect`, but I guess that would already be the case)

simonferquel committed:
Defer error from referencing a non-exisiting context from the config file to 1st API call

When a user removes a context from the store without using the CLI (by removing the context directory directly in the store),
it can trigger an issue where the CLI does not work at all (even for changing the current context) anymore without editing
the config file by hand.

With Docker Desktop and WSL 2, we now link the wsl 2 distros context store to the windows one, so that a context created
from Windows can be used in WSL 2 and vice-versa. However we keep the CLI config files independant (at least for now), so this
issue can be triggered quite often.

This commit fixes that by changing the way we handle context missing error: if the resolved context name comes from the config file
and refers a missing context, the error is reported when the API is actually called instead of at the CLI initialization.
Thus, local-only commands continue to work (shuch as `context use`, `context ls`,...)

Signed-off-by: Simon Ferquel <simon.ferquel@docker.com>

simonferquel committed:
Added validating test

Signed-off-by: Simon Ferquel <simon.ferquel@docker.com>
